docs/advanced/styling.md

```diff
@@ -8,20 +8,337 @@ has_toc: false
+
+Within each component span there is an additional "internal" span, which is the one holding the actual component text.
+The reason for this additional internal span is that it allows CSS styles that closely wrap the text itself, rather than its container box, e.g. for the purpose of drawing a highlight or a box that is exactly in the size of the text.
+
+## Generic and Specific Classes
+
+{: .released }
+Specific classes were introduced in Tasks X.Y.Z.
+
+Each rendered task component (description, priority, recurrence rule etc) includes a **generic class** that denotes this type of component.
+The generic classes are:
+
+- `task-description`
+- `task-priority`
+- `task-due`
+- `task-start`
+- `task-scheduled`
+- `task-done`
+- `task-recurring`
+
+In addition to the generic classes, there are **specific classes** that represent the content of the various task components.
+
+A **priority specific class** represents the specific priority of a class. It can be one of the following:
+
+- `task-priority-high`
+- `task-priority-medium`
+- `task-priority-low`
+- `task-priority-none`
+
+A **date specific class** represents a due, start, scheduled or done date in a format relative to the current date.
+It starts with `task-due-`, `task-start-`, `task-scheduled-` or `task-done-` followed by a relative expression that denotes the number of days this field represents compared to today:
+
+- `task-due-today` (or `task-start-today`, `task-done-today` etc) represents today.
+- `task-due-future-1d` (or `task-start-future-1d`) represents 1 day in the future, i.e. tomorrow.
+- `task-due-past-1d` (or `task-start-past-1d`) represents 1 day in the past, i.e. yesterday.
+- These specific classes are added up to 7 days in the future or past, e.g. `task-scheduled-future-7d` or `task-due-past-7d`.
+- Dates that are further than 7 days in the future or past are given a `far` postfix, e.g. `task-scheduled-future-far` or `task-due-past-far`.
+
+A **tag specific class** translates each task tag into a CSS class, for the purpose of applying formatting according to specific tags.
+It starts with `task-tag-` followed by a *sanitized* version of the tag name, which basically means that characters that are illegal to use in CSS class names are replaced by dashes.
+(CSS class names allow only alphanumeric characters, a dash and an underscore.)
+Examples:
+
+- A task with the tag `#phone` will be added with the specific class `task-tag-phone`.
+- A task with the tag `#t/easy` will be added with the specific class `task-tag-t-easy`.
+- A task with the tag `#task/atHome` will be added the specific class `task-tag-task-atHome`.
+
+Specific classes are added to both their corresponding components (e.g. to the due date component) and also to the complete task `li`, to make it easy for a CSS rule to style a complete task according to some property (e.g. color differently the complete task if it's due today, color a task according to a tag) or just one relevant component.
```

Comment:
I've read this several times and was unsure about it, so did an experiment.

I took this task:

```
- [/] #task Complete this recurring task in **Source view** using **Tasks: Toggle task done** command #context/work #context/laptop 🔁 every day 🛫 2022-02-17 ⏳ 2022-02-18 📅 2022-02-19
```
Inspected its HTML

And then created this snippet:


```css
.task-list-item .task-tag-context-work {
    color: var(--color-red);
}
```

I had thought it would colour the whole line, but it only coloured the description:

![image](https://user-images.githubusercontent.com/4840096/215215448-2f237b5f-9edd-4f7f-9d7a-aeb096ecf542.png)

Out of interest, is there some CSS I could write that would format the whole text of that task as red?

(It may be a feature that only the description is red... I'm just trying to understand)

I then tried formatting the HTML for this, to understand the spans - and got this:

```html
<li class="task-list-item plugin-tasks-list-item task-tag-context-work task-tag-context-laptop task-start-past-far task-scheduled-past-far task-due-past-far is-checked"
    data-task="/" data-line="0">
    <div class="list-bullet"></div>
    <input class="task-list-item-checkbox" type="checkbox" data-line="0">
    <span class="tasks-list-text">
        <span class="task-description task-tag-context-work task-tag-context-laptop">
            <span>Complete this recurring task in <strong>Source view</strong> using <strong>Tasks: Toggle task done</strong> command
                <a href="#context/work" class="tag" target="_blank" rel="noopener">#context/work</a>
                <a href="#context/laptop" class="tag" target="_blank" rel="noopener">#context/laptop</a>
            </span>
        </span>
    <span class="task-recurring">
        <span> 🔁 every day
        </span>
    </span>
    <span class="task-start task-start-past-far">
        <span> 🛫 2022-02-17
        </span>
    </span>
    <span class="task-scheduled task-scheduled-past-far">
        <span> ⏳ 2022-02-18</span>
    </span>
    <span class="task-due task-due-past-far">
        <span> 📅 2022-02-19</span>
        </span>
    </span>
</li>
```

I see that both the tags go on the description. I presume that means that it's not possible to highlight just the text for one of the tags - like make the text of one particular tag stand out - via this mechanism.

(No idea if that matters - just thinking aloud...)

Reply:
> 
> Out of interest, is there some CSS I could write that would format the whole text of that task as red?

I can't currently test this out (might be able tomorrow), but I would assume that this happens because some more specific CSS rules are coloring your task components in black and overriding the red that you set at the parent.
It's interesting to check which rules do that (you can inspect the black components and check the styling applied to them), and possibly override these rules by marking your red rule `!important`.

> 
> I then tried formatting the HTML for this, to understand the spans - and got this:
> 
> ```
> <li class="task-list-item plugin-tasks-list-item task-tag-context-work task-tag-context-laptop task-start-past-far task-scheduled-past-far task-due-past-far is-checked"
>     data-task="/" data-line="0">
>     <div class="list-bullet"></div>
>     <input class="task-list-item-checkbox" type="checkbox" data-line="0">
>     <span class="tasks-list-text">
>         <span class="task-description task-tag-context-work task-tag-context-laptop">
>             <span>Complete this recurring task in <strong>Source view</strong> using <strong>Tasks: Toggle task done</strong> command
>                 <a href="#context/work" class="tag" target="_blank" rel="noopener">#context/work</a>
>                 <a href="#context/laptop" class="tag" target="_blank" rel="noopener">#context/laptop</a>
>             </span>
>         </span>
>     <span class="task-recurring">
>         <span> 🔁 every day
>         </span>
>     </span>
>     <span class="task-start task-start-past-far">
>         <span> 🛫 2022-02-17
>         </span>
>     </span>
>     <span class="task-scheduled task-scheduled-past-far">
>         <span> ⏳ 2022-02-18</span>
>     </span>
>     <span class="task-due task-due-past-far">
>         <span> 📅 2022-02-19</span>
>         </span>
>     </span>
> </li>
> ```
> 
> I see that both the tags go on the description. I presume that means that it's not possible to highlight just the text for one of the tags - like make the text of one particular tag stand out - via this mechanism.
> 
> (No idea if that matters - just thinking aloud...)

I would have liked to support that, but the problem is that in the current way Tasks work, tags are rendered as part as the description using Obsidian's Markdown renderer, which renders them in the way that your snippet shows. From the Tasks renderer perspective, this is one unit of rendered Markdown.
We *can* apply post-processing to Obsidian's rendered markdown and provide this additional functionality, which is definitely useful, I'm just still a little undecided if that's a step too much or a low-hanging fruit for what might be a very popular customization.

Reply:
> I can't currently test this out (might be able tomorrow), but I would assume that this happens because some more specific CSS rules are coloring your task components in black and overriding the red that you set at the parent.
> It's interesting to check which rules do that (you can inspect the black components and check the styling applied to them), and possibly override these rules by marking your red rule `!important`.

Thank you. I tried again, and double-checked that I had no other snippets enabled, and no other themes enabled.

This is the task:

```
- [/] #task Complete this recurring task in **Source view** using **Tasks: Toggle task done** command #context/work #context/laptop 🔁 every day 🛫 2022-02-17 ⏳ 2022-02-18 📅 2022-02-19
```

---

This, as before:


```css
.task-list-item .task-tag-context-work {
    color: var(--color-red);
}
```

Gives:

![image](https://user-images.githubusercontent.com/4840096/215265846-60eb42e3-f2a6-4d1b-8fb8-2c2e59bfb130.png)

This gives the same:

```css
.task-list-item .task-tag-context-work {
    color: var(--color-red) !important;
}
```

---

Changing it to this:

```css
.task-tag-context-work {
    color: var(--color-red);
}
```

Gives this:

![image](https://user-images.githubusercontent.com/4840096/215265875-ba141825-4e6a-4e1a-92ba-7567c8e8d44b.png)

Reply:
I'll investigate where or not this is an expected behavior and possibly make tweaks (even just to documentation).

Reply:
Figured it out. It occurred to me that all those purples were my vault's Accent colour.

If I change the vault's Accent colour to approximately red I get this:

![image](https://user-images.githubusercontent.com/4840096/215268144-20196a78-3e6a-425e-b9fa-af2c3fd33704.png)